### DISCUSSION TITLE
Support `--mode` flag in `astro dev` command

### DISCUSSION THREAD

**Discussion by yacoubb** [withastro/roadmap#870] [Proposal]
### Body

# Summary

Support passing the `--mode` flag when running `astro dev`

# Background & Motivation

Following on from https://github.com/withastro/astro/issues/10509
Vite has the incredibly useful `--mode` flag which allows you to load different `.env.[MODE]` files. This flag is supported by `vite build` and `vite serve`.

`astro dev` does not use the `--mode` flag, `import.meta.env.MODE` is always `development` and only `.env` and `.env.development` are loaded.

Switching between development and staging environments when running astro locally is especially difficult because of this limitation. The best solution at the moment seems to be renaming `.env.development` to swap out the variables needed for development/staging 🤢 
Coming from vite, this seems like a step backwards.

# Goals

- `astro dev` command supports passing `--mode` flag
- `--mode` flag updates `import.meta.env.MODE`
- `--mode` flag controls which `.env.[MODE]` file is loaded

# Example

With an app that logs `import.meta.env.MODE` on startup:

```bash
astro dev
...
> 'development'
```

```bash
astro dev --mode staging
...
> 'staging'
```

**Comment by blikblum** [comment]
I second that feature request.

In my apps i have a script that runs a custom dev build that access remote database instead of local database

I accomplish that with vite setting mode to "remote"

In .env.remote i set a env variable "REMOTE_DATA"

In app i check import.meta.env.REMOTE_DATA to connect to remote data even in dev build

With astro this is not possible currently

Also, the astro docs (https://docs.astro.build/en/guides/environment-variables/) is misleading as it states that can use "any methods supported by vite" to load env variables which is not true


BTW setting env inline in the script command is a poor solution in my case since i develop both in windows and linux

**Comment by nkneupper** [comment]
Definitely support this feature request too. I went down quite a rabbit hole on this only to find out it's not implemented in Astro, especially since the official astro docs state that the `--mode` flag with `astro dev` is possible. https://docs.astro.build/en/guides/environment-variables/#env-files

<img width="739" alt="Screenshot 2025-01-13 at 10 39 29 AM" src="https://github.com/user-attachments/assets/ff9f46ce-3a79-44f5-8d12-7b6baaefc084" />

**Reply by ematipico** [reply]
It is implemented https://docs.astro.build/en/reference/programmatic-reference/#mode

**Comment by ematipico** [comment]
https://docs.astro.build/en/reference/programmatic-reference/#mode

**Reply by kanatbektursyn** [reply]
It appears that `import.meta.env.MODE` within `astro.config.mjs` does not reflect the `--mode` flag when executing `astro dev`, consistently defaulting to `development`. This behavior may be documented, but it seems counter-intuitive.

### Example

When I run the command:

```bash
astro dev --mode staging
```

And the `astro.config.mjs` file contains:

```javascript
// astro.config.mjs
console.log('Environment Mode:', import.meta.env.MODE);
```

The actual output is:

```
Environment Mode: development
```

The expected output was `Environment Mode: staging`.

### Workaround

I have found a workaround by setting the `NODE_ENV` variable in my `package.json` scripts:

```json
// package.json
"scripts": {
  "dev:staging": "NODE_ENV=staging astro dev"
}
```

This allows me to read `process.env.NODE_ENV` in my configuration to determine the environment:

```javascript
// astro.config.mjs
import { loadEnv } from 'vite';
import path from 'path';

const environment = process.env.NODE_ENV; // This correctly becomes 'staging'

// Load variables from /environments/.env.staging
const env = loadEnv(environment, path.resolve(process.cwd(), 'environments'), '');
```

### Desired Behavior

The ideal scenario would be to leverage the `--mode` flag directly within `astro.config.mjs` to dynamically switch between different environment files (e.g., `.env.staging`, `.env.production`) located in a custom `environments/` directory, without relying on setting `NODE_ENV` manually.
